File: examples/hello-world/gabriel-greeting-app-kotlin-compose/app/src/main/kotlin/org/organicprogramming/gabriel/greeting/kotlincompose/settings/SettingsStore.kt
package org.organicprogramming.gabriel.greeting.kotlincompose.settings

import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import org.organicprogramming.gabriel.greeting.kotlincompose.model.CoaxServerTransport
import org.organicprogramming.gabriel.greeting.kotlincompose.model.CoaxSettingsSnapshot
import java.io.File

private const val COAX_ENABLED_KEY = "coax.server.enabled"
private const val COAX_SETTINGS_KEY = "coax.server.settings"
private const val COAX_SERVER_ENABLED_ENV = "OP_COAX_SERVER_ENABLED"
private const val COAX_SERVER_LISTEN_URI_ENV = "OP_COAX_SERVER_LISTEN_URI"

private val prettyGson = GsonBuilder().setPrettyPrinting().create()
private val mapType = object : TypeToken<MutableMap<String, Any?>>() {}.type

interface SettingsStore {
    fun readBool(key: String, defaultValue: Boolean = false): Boolean
    fun readString(key: String, defaultValue: String = ""): String
    fun writeBool(key: String, value: Boolean)
    fun writeString(key: String, value: String)
}

class FileSettingsStore private constructor(
    private val file: File,
    private val values: MutableMap<String, Any?>,
) : SettingsStore {
    override fun readBool(key: String, defaultValue: Boolean): Boolean =
        values[key] as? Boolean ?: defaultValue

    override fun readString(key: String, defaultValue: String): String =
        values[key] as? String ?: defaultValue

    override fun writeBool(key: String, value: Boolean) {
        values[key] = value
        flush()
    }

    override fun writeString(key: String, value: String) {
        values[key] = value
        flush()
    }

    private fun flush() {
        file.writeText(prettyGson.toJson(values) + "\n")
    }

    companion object {
        fun create(): FileSettingsStore {
            val directory = File(settingsDirectoryPath())
            directory.mkdirs()
            val file = File(directory, "settings.json")
            val values = if (file.exists()) {
                runCatching {
                    @Suppress("UNCHECKED_CAST")
                    prettyGson.fromJson<MutableMap<String, Any?>>(file.readText(), mapType) ?: mutableMapOf()
                }.getOrElse { mutableMapOf() }
            } else {
                mutableMapOf()
            }
            return FileSettingsStore(file, values)
        }

        private fun settingsDirectoryPath(): String {
            val home = System.getenv("HOME").orEmpty().trim()
            val appData = System.getenv("APPDATA").orEmpty().trim()
            val xdgConfig = System.getenv("XDG_CONFIG_HOME").orEmpty().trim()
            val osName = System.getProperty("os.name", "").lowercase()

            return when {
                osName.contains("win") && appData.isNotEmpty() ->
                    "$appData/Organic Programming/Gabriel Greeting App Kotlin Compose"
                osName.contains("mac") && home.isNotEmpty() ->
                    "$home/Library/Application Support/Organic Programming/Gabriel Greeting App Kotlin Compose"
                osName.contains("linux") && xdgConfig.isNotEmpty() ->
                    "$xdgConfig/organic-programming/gabriel-greeting-app-kotlin-compose"
                osName.contains("linux") && home.isNotEmpty() ->
                    "$home/.config/organic-programming/gabriel-greeting-app-kotlin-compose"
                else -> "${System.getProperty("user.dir")}/.gabriel-greeting-app-kotlin-compose"
            }
        }
    }
}

class MemorySettingsStore : SettingsStore {
    private val values = mutableMapOf<String, Any?>()

    override fun readBool(key: String, defaultValue: Boolean): Boolean =
        values[key] as? Boolean ?: defaultValue

    override fun readString(key: String, defaultValue: String): String =
        values[key] as? String ?: defaultValue

    override fun writeBool(key: String, value: Boolean) {
        values[key] = value
    }

    override fun writeString(key: String, value: String) {
        values[key] = value
    }
}

fun applyLaunchEnvironmentOverrides(
    store: SettingsStore,
    environment: Map<String, String> = System.getenv(),
) {
    val listenUri = environment[COAX_SERVER_LISTEN_URI_ENV].orEmpty().trim()
    val enabled = parseBoolOverride(environment[COAX_SERVER_ENABLED_ENV])

    if (listenUri.isEmpty() && enabled == null) {
        return
    }

    if (listenUri.isNotEmpty()) {
        store.writeString(COAX_SETTINGS_KEY, snapshotFromListenUri(listenUri).encode())
    }

    when {
        enabled != null -> store.writeBool(COAX_ENABLED_KEY, enabled)
        listenUri.isNotEmpty() -> store.writeBool(COAX_ENABLED_KEY, true)
    }
}

fun coaxEnabledKey(): String = COAX_ENABLED_KEY

fun coaxSettingsKey(): String = COAX_SETTINGS_KEY

private fun parseBoolOverride(value: String?): Boolean? =
    when (value.orEmpty().trim().lowercase()) {
        "1", "true", "yes", "on" -> true
        "0", "false", "no", "off" -> false
        else -> null
    }

private fun snapshotFromListenUri(listenUri: String): CoaxSettingsSnapshot {
    val trimmed = listenUri.trim()
    return when {
        trimmed.startsWith("unix://") -> CoaxSettingsSnapshot(
            serverEnabled = true,
            serverTransport = CoaxServerTransport.UNIX,
            serverHost = CoaxSettingsSnapshot.defaultHost,
            serverPortText = CoaxSettingsSnapshot.defaults.serverPortText,
            serverUnixPath = trimmed.removePrefix("unix://"),
        )
        trimmed.startsWith("tcp://") -> {
            val uri = java.net.URI.create(trimmed)
            val host = uri.host?.trim().orEmpty().ifEmpty { CoaxSettingsSnapshot.defaultHost }
            val port = if (uri.port > 0) uri.port.toString() else CoaxSettingsSnapshot.defaults.serverPortText
            CoaxSettingsSnapshot(
                serverEnabled = true,
                serverTransport = CoaxServerTransport.TCP,
                serverHost = host,
                serverPortText = port,
                serverUnixPath = CoaxSettingsSnapshot.defaultUnixPath,
            )
        }
        else -> CoaxSettingsSnapshot.defaults
    }
}

File: examples/hello-world/gabriel-greeting-app-flutter/app/lib/src/settings_store.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;

import 'model/app_model.dart';

const _coaxEnabledKey = 'coax.server.enabled';
const _coaxSettingsKey = 'coax.server.settings';
const _coaxServerEnabledEnv = 'OP_COAX_SERVER_ENABLED';
const _coaxServerListenUriEnv = 'OP_COAX_SERVER_LISTEN_URI';

abstract interface class SettingsStore {
  bool readBool(String key, {bool defaultValue = false});
  String readString(String key, {String defaultValue = ''});
  Future<void> writeBool(String key, bool value);
  Future<void> writeString(String key, String value);
}

class FileSettingsStore implements SettingsStore {
  FileSettingsStore._(this._file, this._values);

  final File _file;
  final Map<String, Object?> _values;

  static Future<FileSettingsStore> create() async {
    final directory = Directory(_settingsDirectoryPath());
    directory.createSync(recursive: true);

    final file = File(p.join(directory.path, 'settings.json'));
    final values = <String, Object?>{};
    if (file.existsSync()) {
      try {
        final decoded = jsonDecode(await file.readAsString());
        if (decoded is Map) {
          for (final entry in decoded.entries) {
            values[entry.key.toString()] = entry.value;
          }
        }
      } on Object {
        // Recreate the file on next write.
      }
    }

    return FileSettingsStore._(file, values);
  }

  @override
  bool readBool(String key, {bool defaultValue = false}) {
    final value = _values[key];
    return value is bool ? value : defaultValue;
  }

  @override
  String readString(String key, {String defaultValue = ''}) {
    final value = _values[key];
    return value is String ? value : defaultValue;
  }

  @override
  Future<void> writeBool(String key, bool value) async {
    _values[key] = value;
    await _flush();
  }

  @override
  Future<void> writeString(String key, String value) async {
    _values[key] = value;
    await _flush();
  }

  Future<void> _flush() async {
    await _file.writeAsString(
      '${const JsonEncoder.withIndent('  ').convert(_values)}\n',
    );
  }
}

Future<void> applyLaunchEnvironmentOverrides(
  SettingsStore store, {
  Map<String, String>? environment,
}) async {
  final env = environment ?? Platform.environment;
  final listenUri = (env[_coaxServerListenUriEnv] ?? '').trim();
  final enabled = _parseBoolOverride(env[_coaxServerEnabledEnv]);

  if (listenUri.isEmpty && enabled == null) {
    return;
  }

  if (listenUri.isNotEmpty) {
    await store.writeString(
      _coaxSettingsKey,
      _snapshotFromListenUri(listenUri).encode(),
    );
  }

  if (enabled != null) {
    await store.writeBool(_coaxEnabledKey, enabled);
    return;
  }

  if (listenUri.isNotEmpty) {
    await store.writeBool(_coaxEnabledKey, true);
  }
}

String _settingsDirectoryPath() {
  final home = Platform.environment['HOME']?.trim();

  if (Platform.isWindows) {
    final appData = Platform.environment['APPDATA']?.trim();
    if (appData != null && appData.isNotEmpty) {
      return p.join(
        appData,
        'Organic Programming',
        'Gabriel Greeting App Flutter',
      );
    }
  }

  if (Platform.isMacOS && home != null && home.isNotEmpty) {
    return p.join(
      home,
      'Library',
      'Application Support',
      'Organic Programming',
      'Gabriel Greeting App Flutter',
    );
  }

  if (Platform.isLinux) {
    final xdg = Platform.environment['XDG_CONFIG_HOME']?.trim();
    if (xdg != null && xdg.isNotEmpty) {
      return p.join(xdg, 'organic-programming', 'gabriel-greeting-app-flutter');
    }
    if (home != null && home.isNotEmpty) {
      return p.join(
        home,
        '.config',
        'organic-programming',
        'gabriel-greeting-app-flutter',
      );
    }
  }

  return p.join(Directory.current.path, '.gabriel-greeting-app-flutter');
}

class MemorySettingsStore implements SettingsStore {
  final Map<String, Object?> _values = <String, Object?>{};

  @override
  bool readBool(String key, {bool defaultValue = false}) {
    final value = _values[key];
    return value is bool ? value : defaultValue;
  }

  @override
  String readString(String key, {String defaultValue = ''}) {
    final value = _values[key];
    return value is String ? value : defaultValue;
  }

  @override
  Future<void> writeBool(String key, bool value) async {
    _values[key] = value;
  }

  @override
  Future<void> writeString(String key, String value) async {
    _values[key] = value;
  }
}

bool? _parseBoolOverride(String? value) {
  switch ((value ?? '').trim().toLowerCase()) {
    case '1':
    case 'true':
    case 'yes':
    case 'on':
      return true;
    case '0':
    case 'false':
    case 'no':
    case 'off':
      return false;
    default:
      return null;
  }
}

CoaxSettingsSnapshot _snapshotFromListenUri(String listenUri) {
  final trimmed = listenUri.trim();
  if (trimmed.startsWith('unix://')) {
    return CoaxSettingsSnapshot(
      serverTransport: CoaxServerTransport.unix,
      serverHost: CoaxSettingsSnapshot.defaultHost,
      serverPortText: CoaxSettingsSnapshot.defaults.serverPortText,
      serverUnixPath: trimmed.substring('unix://'.length),
    );
  }

  if (trimmed.startsWith('tcp://')) {
    final parsed = Uri.tryParse(trimmed);
    final host = parsed?.host.trim().isNotEmpty == true
        ? parsed!.host.trim()
        : CoaxSettingsSnapshot.defaultHost;
    final port = parsed?.hasPort == true
        ? parsed!.port.toString()
        : CoaxSettingsSnapshot.defaults.serverPortText;
    return CoaxSettingsSnapshot(
      serverTransport: CoaxServerTransport.tcp,
      serverHost: host,
      serverPortText: port,
      serverUnixPath: CoaxSettingsSnapshot.defaultUnixPath,
    );
  }

  return CoaxSettingsSnapshot.defaults;
}
